source/documentation/how-to/ha-vms.html.md

```diff
@@ -0,0 +1,84 @@
+- `KILL`: This is similar to `AUTO_RESUME` unless the VM is paused for too long.  If the storage problem is remedied within a given time interval (80 seconds by default), the VM is automatically resumed.  But if the storage problem persists for longer time, the VM is going to be killed sooner or later.  This behavior is suitable for highly available VMs, in order to permit them to start on another host, which doesn't suffer from the storage problem.
+
+Resume behavior can be selected on any VM, not just highly available VMs.  But there are special considerations regarding highly available VMs:
+
+- As long as a VM is paused, it may prevent Engine from starting it on another, working, host.
+- Or a paused VM may be started on another host and later resumed on the original host, resulting in split brain (the VM runs in multiple instances at the same time).
+
+There is also a special consideration if you use application level high availability in a non-highly available VM.  It may be a bad idea to use `AUTO_RESUME` resume behavior in such a case: If the VM gets paused and the application level high availability handles the situation by running the application elsewhere, confusion may arise if the paused VM starts again.  It may be better to use `LEAVE_PAUSED` or `KILL` resume behaviors in such setups.
+
+## VM leases
+
+VM lease is a feature that guarantees a VM can't be run in multiple instances at the same time.  That guarantee makes restarting highly available VMs safer and thus easier.  Split brain is not possible when a VM has been started with a lease and the lease is kept enabled for the VM.
+
+VM leases can be enabled only for highly available VMs, in High Availability section of the VM editing dialog again.  Just select *Target Storage Domain for VM Lease* in the dialog.
+
+Basically, a VM with a lease can start on a given host only after it acquires its lease there.  Once the VM stops, gets paused, is killed or its QEMU process crashes, it releases the lease, making it available for starting the VM on another host.  VM leases are thoroughly described in a separate [document about leases](/develop/release-management/features/storage/vm-leases/).
+
+To prevent split brain it's best to use VM leases on highly available VMs.  The same also helps restarting highly available VMs on other hosts in case of prolonged storage problems on some of the hosts.  Since paused VMs release their leases, the VMs become available for running on other hosts, but can cause trouble when attempting to resume them later on the original host.  For that reason `KILL` is the only possible resume behavior for VMs with leases, which helps destroy the paused VM instance when Engine moves the VM to Unknown status and attempts to start it elsewhere.  Of course, `KILL` behavior means killing the VM under the given conditions, so think about your setup before using it.
```

Comment:
1. I think this part is incorrect, specifically the "For that reason `KILL` is the only possible resume behavior for VMs with leases". theoretically, VMs should attempt to reacquire their lease when they are resumed from a paused state so it should be safe to use AUTO_RESUME.
2. I think we should emphasize that the `KILL` option may not fit some stateful applications. For instance, let's say that my application holds a state and it takes the application 30 minutes to create this state. In this case, I may prefer to do my best to resume a paused VM that runs this application rather than killing it and starting it elsewhere. There is a trade-off between maximizing the time that application is running (e.g., I may prefer to restart the VM as soon as I find that it can run properly on another host), and reusing its existing state (e.g., I may prefer to wait for a couple of minutes rather than waiting for the VM to boot and 30min for the application to create the state that makes it functional again). A possible solution for this would be not to kill the VM but rather migrate the paused VM to another host where it can be functional with its current state.

Reply:
1. - again, this describes current state, not what it could potentially do. We do not do reacquire because of the auto restart of unknown VMs
2. makes sense to mention the timeouts tradeoff. Though we do not really have a very good estimate of how long it's going to take in all cases

Reply:
1. but I don't think reacquiring the lease is something we should do - libvirt is supposed to do that, do we know that they don't?
2. it depends on the target audience actually - if we want this page to be a technical report of how we consider high availability in oVirt and the mechanism we implemented for that than maybe it is not needed, but if it is targeted for users, that most probably less interested in how we implemented this, but more interested in how they can leverage HA VMs in oVirt in order to make **their application(s)** highly available, then it worth mentioning this trade-off even without concrete numbers IMHO.

Reply:
1. I don't know.

2. Good remark, I added note about that.

Reply:
> but I don't think reacquiring the lease is something we should do - libvirt is supposed to do that, do we know that they don't?

it is reacquired by qemu on resume. But it breaks with our feature to start Unknown VMs blindly (from this host POV). But unless we have more logic on engine side to know that the VM was paused there and we didn't really restart it elsewhere in the meantime we cannot use it.
So again as a reflection of current state this is accurate I believe